### PR TITLE
deprecate keyword color_space for ErrorDiffusion

### PR DESCRIPTION
**Pull request by johnnychen94**
This is not type inferrable and would thus hurt the performance.

Changing the order of parameters is technically breaking, but I doubt people outside are depending on it. Thus an appropriate deprecation of `color_space` keyword should make this a transparent improvement.

### CODE REVIEW

**Comment by johnnychen94** [comment]
I'm surprised that this actually makes things slower...

File: src/error_diffusion.jl [hunk 1]
```diff
@@ -149,12 +152,11 @@ $(_error_diffusion_kwargs)
      Scale."  SID 1975, International Symposium Digest of Technical Papers,
      vol 1975m, pp. 36-37.
 """
-function FloydSteinberg(; kwargs...)
-    return ErrorDiffusion(OffsetMatrix([0 0 7; 3 5 1]//16, 0:1, -1:1); kwargs...)
-end
+FloydSteinberg(CT=XYZ; kwargs...) = ErrorDiffusion{CT}(_floydsteinberg_filter; kwargs...)
+const _floydsteinberg_filter = OffsetMatrix([0 0 7; 3 5 1]//16,0:1,-1:1)
```
Comment (adrhill) [inline src/error_diffusion.jl:156]: This is very nitpicky feedback, but maybe we could call this `const FLOYDSTEINBERG_FILTER` to be consistent with the constant matrices in `ordered.jl`.

File: src/error_diffusion.jl [hunk 2]
```diff
@@ -149,12 +152,11 @@ $(_error_diffusion_kwargs)
      Scale."  SID 1975, International Symposium Digest of Technical Papers,
      vol 1975m, pp. 36-37.
 """
-function FloydSteinberg(; kwargs...)
-    return ErrorDiffusion(OffsetMatrix([0 0 7; 3 5 1]//16, 0:1, -1:1); kwargs...)
-end
+FloydSteinberg(CT=XYZ; kwargs...) = ErrorDiffusion{CT}(_floydsteinberg_filter; kwargs...)
```
[the remark this answers is not in the document]
Comment (adrhill) [reply src/error_diffusion.jl:155]: Master currently defines
```Julia
function ErrorDiffusion(filter; color_space=XYZ, clamp_error=true)
```
on L32. Would changing this to 
```Julia
function ErrorDiffusion(filter, color_space=XYZ; clamp_error=true)
```
be enough for type inference?

**Comment by adrhill** [comment]
> This might be implementation-specific; if the filter weights are `StaticMatrix` then the performance regression here can be eliminated. I'll try to rework the error diffusion dither implementation on top of this PR and see if we get better results.

I've had some shower-thoughts about this. I think we should fully drop the dependency on `OffsetArrays` and use CartesianIndices to iterate over the neighborhoods defined in the filters. This might also simplify a future implementation of #20.

**Comment by adrhill** [comment]
Closing this as it is outdated by #73 , which contains parts of this PR.

**Pull request closed without merging**